### ISSUE TITLE
afterUpdate on create

### ISSUE THREAD

**Issue by RWOverdijk** [balderdashy/waterline#1035]
https://github.com/RWOverdijk/unexpectedUpdate

1. mysql > `create database testcase`
2. npm i
3. sails console
4. The output below (coming from [here](https://github.com/RWOverdijk/unexpectedUpdate/blob/master/api/models/Foo.js#L20)

```bash
sails> Foo.create({bar: {a:'b'}, bat: [{a:'b'}]}).then(console.log).catch(console.error);
{ _bitField: 0,
  _fulfillmentHandler0: undefined,
  _rejectionHandler0: undefined,
  _progressHandler0: undefined,
  _promise0: undefined,
  _receiver0: undefined,
  _settledValue: undefined }
sails> This makes me sad { id: 1,
  createdAt: Fri May 29 2015 16:45:10 GMT+0200 (CEST),
  updatedAt: Fri May 29 2015 16:45:10 GMT+0200 (CEST),
  bar: 1 }
{ bar: 1,
  createdAt: Fri May 29 2015 16:45:10 GMT+0200 (CEST),
  updatedAt: Fri May 29 2015 16:45:10 GMT+0200 (CEST),
  id: 1 }
```

It updates Foo. I'm doing a create. I've tried adding `via`, that's not the issue.

I've traced the update back to [here](https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L73) (which was a nightmare to find).

I don't expect an update at all. Especially not with all values.

On top of that, there's this beauty of a query log:

```sql
INSERT INTO `bar` (`a`, `createdAt`, `updatedAt`) values ('b', '2015-05-29 16:48:57', '2015-05-29 16:48:57')
INSERT INTO `foo` (`bar`, `createdAt`, `updatedAt`) values (1, '2015-05-29 16:48:57', '2015-05-29 16:48:57')
SELECT `foo`.`id`, `foo`.`createdAt`, `foo`.`updatedAt`, `foo`.`bar` FROM `foo` AS `foo`  WHERE `foo`.`id` = 1
UPDATE `foo` AS `foo` SET `bar` = 1, `createdAt` = '2015-05-29 16:48:57', `updatedAt` = '2015-05-29 16:48:57', `id` = 1  WHERE `foo`.`id` = 1
SELECT `foo`.`id`, `foo`.`createdAt`, `foo`.`updatedAt`, `foo`.`bar` FROM `foo` AS `foo`  WHERE `foo`.`id` = 1  LIMIT 1
INSERT INTO `bat` (`a`, `foo`, `createdAt`, `updatedAt`) values ('b', 1, '2015-05-29 16:48:57', '2015-05-29 16:48:57')
SELECT `foo`.`id`, `foo`.`createdAt`, `foo`.`updatedAt`, `foo`.`bar`, `__bar`.`a` AS `bar___a`, `__bar`.`id` AS `bar___id`, `__bar`.`createdAt` AS `bar___createdAt`, `__bar`.`updatedAt` AS `bar___updatedAt`, `__bar`.`foo` AS `bar___foo` FROM `foo` AS `foo`  LEFT OUTER JOIN `bar` AS `__bar` ON `foo`.`bar` = `__bar`.`id` WHERE `foo`.`id` = 1
(SELECT * FROM `bat` AS `bat` WHERE `foo` = 1  ORDER BY `bat`.`id` ASC)
```

So, insert good. Another... What? Order: Insert, select, update, select.

I think something's wrong here.

This is all latest tag install.

**Comment by devinivy** [comment]
Are you using nested items, or are these `json` type fields?  If you are using nested items, do you experience the same thing when doing a `create` without nested items?

**Comment by dmarcelino** [comment]
This does look like a bug, I can't see any reason for a `.create` with only nested creates to be doing update (and select/find) operations.

**Comment by devinivy** [comment]
Nested updates/creates need some love.  I expect the update you're seeing is caused by the nested create routine, specifically this line: https://github.com/balderdashy/waterline/blob/master/lib/waterline/utils/nestedOperations/create.js#L62 .

**Comment by particlebanana** [comment]
Looks like an extra `.save` on the belongsTo association as pointed out by @devinivy. The extra finds are due to the way mysql returns values. It only returns the id of the record(s) updated so in order to send back the complete record to match the api an extra find is needed.

**Comment by RWOverdijk** [comment]
@particlebanana Okay. So how can it be fixed? Also, what extra ID are you talking about? After a create you get an ID. There' s no need for an update. Or do you mean that the extra find would be solved with this, because there's no update anymore?

**Comment by dmarcelino** [comment]
> Nested updates/creates need some love. 

Pardon me for the following (slight) off-topic but there are a few issues concerning nested updates/creates that probably should be tackled together (probably in `0.11`):
* this #1035
* model.save() fails when model has been populated #975
* Make m:n nested updates API consistent with 1:1 and 1:m #965
* beforeValidate called 3 times on update with association #751
* Associations not reduced on update #646
* Model.save() method should not always populate associations #441
* Add `.add()`, `.remove()`, `.save()` support for 1-1 and 1-* associations #292

These are all I can remember now and I've finally grouped them in the same place. Now we can go back to discussing the issue at hand.

**Comment by RWOverdijk** [comment]
I'm sort of on a tight deadline here so any pointers to get this fixed would be appreciated.

**Comment by devinivy** [comment]
Given that using `save` is the de-facto way to add/remove items to/from collections, and `save` invariably performs an update, I think you need to start there.  One thing we've wanted for a while is for `save` to diff changes and only perform the necessary updates.  In this case, `save` is used inside the nested update routine to add `bat` records with correctly-associated `foo` records.  If `save` used diffing, it would know that there doesn't need to be any update to the parent `foo` record.

**Comment by RWOverdijk** [comment]
But in order to diff, you'll still need the extra find. Which is a hack and still only fixes half of the problem.

**Comment by devinivy** [comment]
There shouldn't be any extra `find`s beyond what already happens to achieve diffing.  You only need to diff changes to the model instance to see which properties have been modified.  With diffing, you'll nix one waterline update call, which consists of an UPDATE and a SELECT mysql query.

**Comment by dmarcelino** [comment]
There is already a PR for detecting changes: PR #710, reviewing that could be a good starting point.

**Comment by RWOverdijk** [comment]
Commenting out the last "parentModel.save(cb);" "fixes" it. Why is this call here?

**Comment by devinivy** [comment]
Removing that should break `create` with nested records in a many-to-many relationship.  Saving that record creates all the associations described in the rest of that file.  I think your frustration is partially around duplicate functionality between [createBelongsTo](https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/create.js#L116) and [nestedOperations](https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/create.js#L227).

**Comment by RWOverdijk** [comment]
Following that logic... Shouldn't it be possible to detect if it's a m2m relation and only then call that create?

**Comment by dmarcelino** [comment]
PR #1048 fixes this.

**Comment by sailsbot** [comment]
Thanks for posting, @RWOverdijk.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
